### ISSUE TITLE
Bring in camera code donation from Dell

### ISSUE THREAD

**Issue by alice-sowerby** [FreeBSDFoundation/proj-laptop#39]
## User Story
As a laptop user I want to be able to use the camera on my laptop as it will make video calls etc much better.

## Implementation Notes
Work with Dell to integrate the code donation for laptop support with the FreeBSD source code. The code is written but some portions are being rewritten for licence purposes. This is a work in progress. 

## Acceptance Criteria
- [ ] 

## Discussion thread
https://lists.freebsd.org/archives/freebsd-desktop/2025-March/005617.html

**Comment by lwhsu** [comment]
> `webcamd` already does that? provides V4L layer and many more features?

Yes webcamd works nicely. The one Dell has is an implementation purely in kernel. Please see https://www.freebsd.org/status/report-2024-07-2024-09/#_freebsd_v4l2_kernel_usb_video_class_driver for more information.

**Comment by lwhsu** [comment]
2024Q4 report has been published: https://www.freebsd.org/status/report-2024-10-2024-12/#_freebsd_v4l2_kernel_usb_video_class_driver

The code is being review and adjust for merging.

**Comment by lwhsu** [comment]
> `webcamd` already does that? provides V4L layer and many more features?

It's a kernel implementation and aim for providing more modern features.

**Comment by alice-sowerby** [comment]
@lwhsu do you have any updates for March?

**Comment by alice-sowerby** [comment]
@lwhsu do you have any updates for April?

**Comment by lwhsu** [comment]
> [@lwhsu](https://github.com/lwhsu) do you have any updates for April?

I'm still testing and reviewing the code from Dell. Aiming to be merged in the mid of June.

**Comment by alice-sowerby** [comment]
@lwhsu do you have any updates for May?

**Comment by alice-sowerby** [comment]
The team at Dell that was helping with the work has been dissolved, and we now have no contact at Dell to support the effort. We will review the options of how to progress on this.

**Comment by lwhsu** [comment]
Still working on reorganizing it but I got some good progress last month. I have separated the code to uvc(4) and video(4) parts. Need more cleanup and review the architecture.

I'll try to finish that in Q2.

**Comment by lwhsu** [comment]
**Status update (July 2026)**

A native `uvideo(4)` driver (developed independently, not sponsored by this project) has been merged into FreeBSD `main`. This is a good base for us to build on, and it also changes our work: we can do less in some parts, but need to do more in others.

The original plan was mainly cleaning up and integrating the donated code. Since `uvideo(4)` is now in `main`, we adjusted the approach: use `main` as the only baseline, and bring in the donated work as smaller, independently reviewable patches on top of it, instead of maintaining a parallel implementation. We evaluated this and believe it is a better way than the original plan.

Currently we are doing a more detailed comparison of the architecture and features of `uvideo(4)`, the donated `uvc(4)` code, `webcamd`, and other open source implementations (Linux, OpenBSD/NetBSD, etc.), and writing a more detailed technical plan based on that. The near-term goal is stabilizing basic UVC capture on the in-tree driver: streaming (read/mmap), controls, clean hot-unplug, with regression tests, so common webcams work reliably. Then application support (browsers, FFmpeg, etc.) on the native path. `webcamd` stays as a fallback for devices not yet covered.
